src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -398,30 +507,26 @@ private bool TryGetPropertyForInputModelProperty(Func<ObjectTypeProperty, bool>
 
         protected override ObjectTypeDiscriminator? BuildDiscriminator()
         {
-            string? discriminatorPropertyName = _inputModel.DiscriminatorPropertyName;
-            ObjectTypeDiscriminatorImplementation[] implementations = Array.Empty<ObjectTypeDiscriminatorImplementation>();
-            Constant? value = null;
-            ObjectTypeProperty property;
+            var parentDiscriminator = GetBaseObjectType()?.Discriminator;
+            var property = Properties.FirstOrDefault(p => p.InputModelProperty is not null && p.InputModelProperty.IsDiscriminator)
+                ?? parentDiscriminator?.Property;
 
-            if (discriminatorPropertyName == null)
-            {
-                var parent = GetBaseObjectType();
-                if (parent is null || parent.Discriminator is null)
-                {
-                    //neither me nor my parent are discriminators so I can bail
-                    return null;
-                }
+            var discriminatorPropertyName = _inputModel.DiscriminatorPropertyName ?? parentDiscriminator?.SerializedName;
 
-                discriminatorPropertyName = parent.Discriminator.SerializedName;
-                property = parent.Discriminator.Property;
-            }
-            else
+            //neither me nor my parent are discriminators so I can bail
+            if (property is null || discriminatorPropertyName is null)
             {
-                //only load implementations for the base type
-                implementations = _derivedTypes!.Select(child => new ObjectTypeDiscriminatorImplementation(child.DiscriminatorValue!, _typeFactory.CreateType(child))).ToArray();
-                property = Properties.First(p => p.InputModelProperty is not null && p.InputModelProperty.IsDiscriminator);
+                return null;
             }
 
+            Constant? value = null;
+
+            //only load implementations for the base type
+            var implementations = Configuration.Generation1ConvenienceClient
+                ? GetDerivedTypes(_derivedTypes).OrderBy(i => i.Key).ToArray()
```

Comment:
is this `orderBy` call because of reducing the amount of generated code changes?

Reply:
Yes. Added a TODO.